x-pack/metricbeat/module/googlecloud/stackdriver/metricset.go

```diff
@@ -0,0 +1,146 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package stackdriver
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
+
+	"github.com/elastic/beats/libbeat/common"
+
+	"google.golang.org/api/option"
+
+	"github.com/elastic/beats/x-pack/metricbeat/module/googlecloud"
+
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/metricbeat/mb"
+)
+
+const (
+	METRICSET_NAME = "stackdriver"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet(googlecloud.MODULE_NAME, METRICSET_NAME, New)
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	config config
+}
+
+type config struct {
+	Metrics             []string `config:"stackdriver.metrics" validate:"required"`
+	Zone                string   `config:"zone" validate:"required"`
+	ProjectID           string   `config:"project_id" validate:"required"`
+	IncludeLabels       bool     `config:"include_labels"`
+	ServiceName         string   `config:"stackdriver.service"  validate:"required"`
+	CredentialsFilePath string   `config:"credentials_file_path"`
+
+	opt option.ClientOption
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The gcp '%s' metricset is beta.", METRICSET_NAME)
+
+	m := &MetricSet{BaseMetricSet: base}
+
+	if err := base.Module().UnpackConfig(&m.config); err != nil {
+		return nil, err
+	}
+
+	m.config.opt = option.WithCredentialsFile(m.config.CredentialsFilePath)
+
+	return m, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+	err := validatePeriodForGCP(m.Module().Config().Period)
+	if err != nil {
+		return errors.Wrap(err, "invalid time period for GCP")
+	}
+
+	//TODO Add credentials check?
```

Comment:
I removed the TODO finally. I didn't require credentials because a user executing Metricbeat inside a GCP machine can already authenticate itself using the machine, so credentials of any kind aren't needed.

At the same time, a user without credentials out of GCP will receive an error in console in case the authentication failed.